src/operators.jl

```diff
@@ -858,3 +858,191 @@ function merge_vertices!(g::Graph{T}, vs::Vector{U} where U <: Integer) where T
+
+Reference -
+Lehot, Philippe G. H. (1974), "An optimal algorithm to detect a line graph and output its root graph"
+"""
+function get_root(g::SimpleGraph{Int})
+    fnn = Dict()
+    hnn = Dict()
+    in_fnn = falses(nv(g))
+    in_hnn = falses(nv(g))
+    curnt_clq = Array{Int, 1}()
+
+    #first step of algorithm
+    first_edge = first(edges(g))
+    basic_1 = first_edge.src
+    basic_2 = first_edge.dst
+    fnn[basic_1] = [1, 2]
+    fnn[basic_2] = [2, 3]
+    common = common_neighbors(g, basic_1, basic_2)
+    ncom = length(common)
+    N = 4
+    if ncom == 0
+        #do nothing
+    elseif ncom == 1
+        x = common[1]
+        if !is_evenTri(g, (basic_1, basic_2, x))
+            fnn[x] = [2, 4]
+            N += 1
+        end
+    elseif ncom == 2
+        x, y = common
+        if has_edge(g, x, y)
+            fnn[x] = [2, 4]
+            fnn[y] = [2, 5]
+            N += 2
+        else
+            xeven = is_evenTri(g, (basic_1, basic_2, x))
+            yeven = is_evenTri(g, (basic_1, basic_2, y))
+            if !xeven || !yeven
+                if !xeven
+                    fnn[x] = [2, 4]
+                else
+                    fnn[y] = [2, 4]
+                end
+            else
+                fnn[y] = [2, 4]
+            end
+            N += 1
+        end
+    else    #ncom >=3
+        a, b, c = common
+        if !has_edge(g, a, b)
+            if has_edge(g, a, c)
+                fnn[a] = [2, 4]
+            else
+                fnn[b] = [2, 4]
+            end
+        else
+            fnn[a] = [2, 4]
+        end
+        N += 1
+        for u ∈ common
+            u ∈ keys(fnn) && continue
+            is_incliq = true
+            for v ∈ keys(fnn)
+                if !has_edge(g, u, v)
+                    is_incliq = false
+                    break
+                end
+            end
+            !is_incliq && continue
+            fnn[u] = [2, N]
+            N += 1
+        end
+    end
+    for u ∈ keys(fnn)
+        push!(curnt_clq, u)
+        in_fnn[u] = true
+    end
+    name_adj(curnt_clq, fnn, hnn, 2, in_hnn, in_fnn, g)
+
+    #main step of algorithm
+    while !isempty(hnn)
+        key,val = first(hnn)
+        hn_node = key
+        cliqn,fn_adjn = val
+        fnn[key] = [cliqn, N]
+        in_fnn[key] = true
+        delete!(hnn, key)
+        in_hnn[key] = false
+        push!(curnt_clq, key)
+        N += 1
+        common = common_neighbors(g, hn_node, fn_adjn)
+        for v ∈ common
+            in_fnn[v] && continue
+            fnn[v] = [cliqn, N]
+            in_fnn[v] = true
+            delete!(hnn, v)
+            in_hnn[v] = false
+            push!(curnt_clq, v)
+            N += 1
+        end
+        name_adj(curnt_clq, fnn, hnn, cliqn, in_hnn, in_fnn, g)
+    end
+    root_graph = SimpleGraph{Int}(N-1)
+    for val in values(fnn)
+        add_edge!(root_graph, val[1], val[2])
+    end
+    if !LightGraphs.Experimental.has_isomorph(g, linegraph(root_graph))
```

Comment:
Isn't there a better way to do that? Isomorphism testing can be very expensive.

Reply:
A better way is to remove the isomorphism check and create another function that checks for valid linegraph.If the user calls the get_root function and gets an incorrect output then it is their fault.

Reply:
Hmm depends how difficult that check is, maybe it can also be an optional flag to the function.
In any case, I just looked at Lehot's paper, and I think it should be possible to verify that a graph has a root without a possibly exponential isomorphism check.

Reply:
I might have misunderstood the paper then, but could you give a general idea of how it would be done ?

Reply:
For that I would need to study the paper more and I first need to time for that. But I think you can use the labeling that is created when generating the root graph, if I understood that part correctly.

The paper also cites another result for checking if the graph is a line graph in the beginning but I think using that result directly would lead to an algorithm of complexity roughly O(n^5)

Reply:
If I recall correctly, we generate the root graph first and then obtain its line graph. Then we check if the edges in the generated line graph are the same as the original line graph. But the problem here is that due labeling of the root graph and then again labeling of its line graph, will lead to a new labeling altogether, that's why I used the isomorphism check.

Reply:
Btw what is the complexity of isomorphism check ?

Reply:
Seems exponential

Reply:
Okay I thought of an approach making use of the labeling, will test it out.